src/graph.ts

```diff
@@ -11,35 +11,148 @@
+ *
+ * @param cfnType - The CloudFormation type (i.e. AWS::S3::Bucket)
+ * @returns The resource type (i.e. Bucket)
+ */
+function typeFromCfn(cfnType: string): string {
+    const typeParts = cfnType.split('::');
+    if (typeParts.length !== 3) {
+        throw new Error(`Expected cfn type in format 'AWS::Service::Resource', got ${cfnType}`);
+    }
+    return typeParts[2];
+}
+
+function typeFromFqn(fqn: string): string {
+    const fqnParts = fqn.split('.');
+    return fqnParts[fqnParts.length - 1];
 }
 
 export class GraphBuilder {
-    constructNodes: Map<Construct, GraphNode>;
+    // Allows for easy access to the GraphNode of a specific Construct
+    constructNodes: Map<ConstructInfo, GraphNode>;
+    // Map of resource logicalId to GraphNode. Allows for easy lookup by logicalId
     cfnElementNodes: Map<string, GraphNode>;
 
-    constructor(private readonly stack: Stack) {
-        this.constructNodes = new Map<Construct, GraphNode>();
+    constructor(private readonly stack: StackManifest) {
+        this.constructNodes = new Map<ConstructInfo, GraphNode>();
         this.cfnElementNodes = new Map<string, GraphNode>();
     }
 
     // build constructs a dependency graph from the adapter and returns its nodes sorted in topological order.
-    public static build(stack: Stack): GraphNode[] {
+    public static build(stack: StackManifest): GraphNode[] {
         const b = new GraphBuilder(stack);
         return b._build();
     }
 
+    /**
+     * Recursively parses the construct tree to create:
+     * - constructNodes
+     * - cfnElementNodes
+     *
+     * @param tree - The construct tree of the current construct being parsed
+     * @param parent - The parent construct of the construct currently being parsed
+     */
+    private parseTree(tree: ConstructTree, parent?: ConstructInfo) {
+        const construct: ConstructInfo = {
+            parent,
+            id: tree.id,
+            path: tree.path,
+            type: tree.constructInfo ? typeFromFqn(tree.constructInfo.fqn) : tree.id,
+            attributes: tree.attributes,
+        };
+        const node: GraphNode = {
+            incomingEdges: new Set<GraphNode>(),
+            outgoingEdges: new Set<GraphNode>(),
+            construct,
+        };
+        if (tree.attributes && 'aws:cdk:cloudformation:type' in tree.attributes) {
+            const cfnType = tree.attributes['aws:cdk:cloudformation:type'] as string;
+            const logicalId = this.stack.logicalIdForPath(tree.path);
+            const resource = this.stack.resourceWithLogicalId(logicalId);
+            const typ = typeFromCfn(cfnType);
+            node.construct.type = typ;
+            construct.type = typ;
+            if (resource.Type === cfnType) {
+                node.resource = resource;
+                node.logicalId = logicalId;
+                this.cfnElementNodes.set(logicalId, node);
+            } else {
+                console.error('SOMETHING WENT WRONG: ', tree);
```

Comment:
I understand this is temporary, but do we want console.error or do we want throw new Error (fail fast)?

Reply:
Updated!